### PR TITLE
Add integrated gradients example

### PR DESCRIPTION
**Pull request by AakashKumarNain**

### CODE REVIEW

File: examples/vision/integrated_gradients.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+one of the interpretibility techniques used to understand why did the model make
+this prediction on this input.
+
+[Integrated Gradients](https://arxiv.org/abs/1701.07875) involves following steps:
+1. Construct a sequence of images interpolating from a baseline to the actual image.
+2. Do a forward pass on these interpolated images and get the gradients.
+2. Average the gradients across these images.
+
+Refrences:
+Paper: https://arxiv.org/abs/1701.07875
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:22]: Please use md formatting for the links.

File: examples/vision/integrated_gradients.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+"""
+
+"""
+## Integrated Gradients
+[Integrated Gradients](https://arxiv.org/abs/1701.07875) is a technique for
+attributing a classification model's prediction to its input features. It is
+one of the interpretibility techniques used to understand why did the model make
+this prediction on this input.
+
+[Integrated Gradients](https://arxiv.org/abs/1701.07875) involves following steps:
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:16]: These steps may not be very clear for someone not already familiar with the algorithm. Please improve clarity.

File: examples/vision/integrated_gradients.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+Title: Integrated Gradients
+Author: [A_K_Nain](https://twitter.com/A_K_Nain)
+Date created: 2020/06/2
+Last modified: 2020/06/2
+Description: How to obtain integrated gradients for a classification model.
+"""
+
+"""
+## Integrated Gradients
+[Integrated Gradients](https://arxiv.org/abs/1701.07875) is a technique for
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:11]: is a model interpretability technique: you can use it to visualize the relationship between input features and model predictions.

File: examples/vision/integrated_gradients.py [hunk 4]
```diff
@@ -0,0 +1,474 @@
+"""
+Title: Integrated Gradients
+Author: [A_K_Nain](https://twitter.com/A_K_Nain)
+Date created: 2020/06/2
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:4]: 02

File: examples/vision/integrated_gradients.py [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy import ndimage
+from pathlib import Path
+from IPython.display import Image
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from tensorflow.keras.applications.xception import (
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:40]: In this case, you can simply import `applications.xception` then access the different objects from the module.

File: examples/vision/integrated_gradients.py [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+
+
+def get_gradients(img_input, top_pred_idx):
+    """Computes the gradients of outputs w.r.t input image.
+
+    Args:
+        img_input: 4D image tensor
+        top_pred_idx: predicted label for the input image
+    Returns:
+        gradients of the predictions w.r.t to img_input
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:76]: "to" is redundant

File: examples/vision/integrated_gradients.py [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+    return array
+
+
+def get_gradients(img_input, top_pred_idx):
+    """Computes the gradients of outputs w.r.t input image.
+
+    Args:
+        img_input: 4D image tensor
+        top_pred_idx: predicted label for the input image
+    Returns:
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:75]: Add line break above

File: examples/vision/integrated_gradients.py [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+        gradients of the predictions w.r.t to img_input
+    """
+    images = tf.cast(img_input, tf.float32)
+
+    # use gradient tape for computing gradients
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:80]: In general, capitalize all comments. This specific comment can be removed since it does not add more information compared to the code itself.

File: examples/vision/integrated_gradients.py [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+
+    # 1. Do interpolation
+    img_input = img_input.astype(np.float32)
+    interpolated_image = [
+        baseline + (step / num_steps) * (img_input - baseline)
+        for step in range(num_steps + 1)
+    ]
+    interpolated_image = np.array(interpolated_image).astype(np.float32)
+
+    # 2. Preproess the interpolated images
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:116]: Please proofread for typos

File: examples/vision/integrated_gradients.py [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+    return tf.reduce_mean(integrated_grads, axis=0)
+
+
+"""
+## Helper class for visualizing gradients and integrated gradients
+"""
+
+
+class GradVisualizer:
+    """Plot gradients of an input."""
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:176]: w.r.t. ...

File: examples/vision/integrated_gradients.py [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+            clip_above_percentile=clip_above_percentile,
+            clip_below_percentile=clip_below_percentile,
+            morphological_cleanup=morphological_cleanup,
+            structure=structure,
+            outlines=outlines,
+            outlines_component_percentage=outlines_component_percentage,
+            overlay=overlay,
+        )
+
+        f, ax = plt.subplots(1, 3, figsize=figsize)
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:417]: You can use `_` for variables you don't use, e.g. `_, ax = `

File: examples/vision/integrated_gradients.py [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+        sorted_sums_and_masks = sorted(component_sums, key=lambda x: x[0], reverse=True)
+        sorted_sums = list(zip(*sorted_sums_and_masks))[0]
+        cumulative_sorted_sums = np.cumsum(sorted_sums)
+        cutoff_threshold = percentage * total / 100
+        cutoff_idx = np.where(cumulative_sorted_sums >= cutoff_threshold)[0][0]
+
+        if cutoff_idx > 2:
+            cutoff_idx = 2
+
+        # 6.Set the values for the kept components.
```
Comment (fchollet) [inline examples/vision/integrated_gradients.py:291]: The comments are good. Make sure to proofread for inconsistencies and typos (e.g. lack of space after period above, double period on line 363, use of comma instead of period, etc)

**Comment by AakashKumarNain** [comment]
Thanks @fchollet for the review. I will make the necessary changes

**Comment by AakashKumarNain** [comment]
Thanks for the review. I had added the autogenerated files

**Comment by fchollet** [comment]
Thanks a lot for the great contribution!
